### PR TITLE
update fink for macOS Mojave 10.14

### PR DESCRIPTION
**Pull request by jwhowarth**
update fink for macOS Mojave 10.14

### CODE REVIEW

**Comment by jwhowarth** [comment]
Filed radar://43963726 "macOS_SDK_headers_for_macOS_10.14.pkg fails to install perl headers" about the fact that builds with makemaker on 10.14 are failing due to its invocation of -iwithsysroot "/System/Library/Perl/5.18/darwin-thread-multi-2level/CORE" which only contains libperl.dylib but no headers like both /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Perl/5.18/darwin-thread-multi-2level/CORE and /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Perl/5.18/darwin-thread-multi-2level/CORE

**Comment by jwhowarth** [comment]
Passed the following change on to Apple by radar which would fix the issue..

`
--- /System/Library/Perl/5.18/ExtUtils/MM_Unix.pm.orig  2018-09-01 20:46:43.000000000 -0400
+++ /System/Library/Perl/5.18/ExtUtils/MM_Unix.pm       2018-09-01 20:40:09.000000000 -0400
@@ -131,7 +131,7 @@
     my(@m);
     
     my $command = '$(CCCMD)';
-    my $flags   = '$(CCCDLFLAGS) -iwithsysroot "$(PERL_INC)" $(PASTHRU_DEFINE) $(DEFINE)';
+    my $flags   = '$(CCCDLFLAGS) -iwithsysroot "$(PERL_SYSROOT)$(PERL_INC)" $(PASTHRU_DEFINE) $(DEFINE)';
     
     if (my $cpp = $Config{cpprun}) {
         my $cpp_cmd = $self->const_cccmd;
@@ -3595,7 +3595,7 @@
     '
 .xs$(OBJ_EXT):
        $(XSUBPPRUN) $(XSPROTOARG) $(XSUBPPARGS) $*.xs > $*.xsc && $(MV) $*.xsc $*.c
-       $(CCCMD) $(CCCDLFLAGS) -iwithsysroot "$(PERL_INC)" $(PASTHRU_DEFINE) $(DEFINE) $*.c
+       $(CCCMD) $(CCCDLFLAGS) -iwithsysroot "$(PERL_SYSROOT)$(PERL_INC)" $(PASTHRU_DEFINE) $(DEFINE) $*.c
 ';
 }
`

**Comment by jwhowarth** [comment]
A few more notes on building using Xcode 10 under 10.14. 

1) The building of i386 binaries is prohibited under the 10.14 SDK. They achieve this by having the /Developer/SDKs/MacOSX.sdk/usr/lib/libSystem.B.tbd only expose the x86_64 code from /usr/lib/libSystem.dylib. 

cc -v -arch i386 -arch x86_64  -isysroot /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -mmacosx-version-min=10.14 -Wl,-search_paths_first -Wl,-headerpad_max_install_names himenoBMTxpa.c
...
ld: symbol(s) not found for architecture i386

So you get the oddity that i386 code can still be built as long as you don't use the SDK itself.

2) Upstream really wants developers to avoid installing the fall back header package. I don't quite understand the nuances yet but it does appear that the perl header glitch doesn't exist as long as the macOS_SDK_headers_for_macOS_10.14.pkg is not installed. When it is installed, the absence of the Perl headers in /System/Library/Perl/5.18/darwin-thread-multi-2level/CORE will break perl module builds. However, not installing the fall back header package means that fink needs to relax the restriction that dev-tools with a populated /usr/include must be installed. I'll look into what additional patches to fink would be required to allow that on 10.14 and how successful Xcode 10 is at bootstrapping and building under fink in the absence of the fall back header package.

**Comment by jwhowarth** [comment]
For builds like llvm which select the 10.14 SDK using -isysroot, one can hack around the obsoleting of i386 linkages by also passing --sysroot=/ for these problematic linkages so that libSystem.dylib is used instead of the tbd.

**Comment by jwhowarth** [comment]
Relaxed test for /usr/include on darwin18 and later to allow for the SDK to provide the headers rather than forcing the fall back header package to be installed (which Apple discourages).

**Comment by jwhowarth** [comment]
Just to be clear, from the Xcode 10 beta release notes...

Command Line Tools
The Command Line Tools package installs the macOS system headers inside the macOS SDK.
Software that compiles with the installed tools will search for headers within the macOS SDK
provided by either Xcode at:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/
SDKs/MacOSX10.14.sdk
or the Command Line Tools at:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
depending on which is selected via xcode-select.
The command line tools will search the SDK for system headers by default. However, some software
may fail to build correctly against the SDK and require macOS headers to be installed in the base
system under /usr/include. If you are the maintainer of such software, we encourage you to
update your project to work with the SDK or file a bug report for issues that are preventing you from
doing so. As a workaround, an extra package is provided which will install the headers to the base
system. In a future release, this package will no longer be provided. You can find this package at/Library/Developer/CommandLineTools/Packages/macOS_SDK_headers_for_macOS_10.14.
pkg
To make sure that you are using the intended version of the command line tools, run xcode-select
-s <path to Xcode> or xcode select -s /Library/Developer/CommandLineTools after
installing. (38476941)

**Comment by jwhowarth** [comment]
The biggest issue in building without the root level /usr/include will likely not come from the average package build that just uses the Xcode 10 clang compilers but rather those using FSF gcc and llvm.org's clang compiler which both lack the changes to automatically add the SDK header paths into the search path. Fortunately gfortran doesn't use anything in /usr/include so the only immediate issue is fixing the 3-stage FSF gcc bootstrap to find the SDK headers in stage2/stage3 so that it builds in the absence of /usr/include.

**Comment by jwhowarth** [comment]
There should be no reason to pass additional flags on the compiler wrappers for clang as the Xcode 10 compiler already automatically appends...

-resource-dir /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/10.0.0 -isysroot /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -I/usr/local/include

for both simple compiles and for linkages. The problem compilers are those outside of Xcode 10 such as llvm.org and FSF gcc which need to become SDK -isysroot savvy.

**Comment by jwhowarth** [comment]
Note that the proposed changes (without the help perl find its headers on 10.14 hunk) successfully bootstraps fink trunk on 10.14 with the Command Line Tools installed without the fall back header package.

**Comment by jwhowarth** [comment]
I have a clarification from Jeremy on how the SDK is found in the Xcode 10 clang. It is done by the compiler wrappers in /usr/bin passing the call to the buried compiler through xcrun which sets the location of the SDK. You can see this from the fact that, in the absence of the fall back header package, compiling with the buried clang compiler at /Library/Developer/CommandLineTools/usr/bin/clang fails with it being unable to find the /usr/include headers. The fact that is done with xcrun is also confirmed with...

$ grep xcrun /usr/bin/clang
Binary file /usr/bin/clang matches

**Comment by jwhowarth** [comment]
It appears that Apple has some work to do in order to make the system perl SDK friendly. The issue is that MakeMaker looks at  /System/Library/Perl/5.18/darwin-thread-multi-2level/Config.pm which hardcodes in / level paths. This results in the builds using MakeMaker (including its own test suite) always wanting the perl headers to reside in / based paths rather than from the SDK.

**Comment by jwhowarth** [comment]
Note that the Xcode 10 GM release is still missing the Perl headers in the macOS_SDK_headers_for_macOS_10.14.pkg installed by the Command Line Tools package. So some hack like the one proposed...

[https://github.com/fink/fink/pull/166/commits/a3424c4ed5d44820579a5edd7505e908d4e02bd6](https://github.com/fink/fink/pull/166/commits/a3424c4ed5d44820579a5edd7505e908d4e02bd6)

will be required to avoid fink breakage when the fall back SDK header package is installed.

File: perlmod/Fink/Bootstrap.pm [hunk 1]
```diff
@@ -245,6 +245,12 @@ GCC_MSG
 			"of Mac OS X might work with Fink, but there are no " .
 			"guarantees.");
 		$distribution = "10.13";
+	} elsif ($host =~ /^i386-apple-darwin(\d+)\./) {
```
Comment (dmacks) [inline perlmod/Fink/Bootstrap.pm]: This should probably be a pair of "darwin18" checks, not a duplicate of the next block that is the fallback for unknown kernel numbers.

**Comment by jwhowarth** [comment]
The Xcode 10.1 beta 2 Command Line Tools for 10.14 now contains the previously missing Perl headers in the installed fall back SDK package. So it may be easier to just hold off supporting 10.14 until Xcode 10.1 is released and blacklist Xcode 10.0 as an unsupported version due to that bug.

File: perlmod/Fink/Bootstrap.pm (hunk 1 above)
Comment (remkos) [reply perlmod/Fink/Bootstrap.pm]: You are right @dmacks. There should indeed be a pair of darwin18 checks as above.

**Comment by remkos** [comment]
@jwhowarth, can you please make the change suggested by @dmacks above on the "darwin18 checks". The code will work as it is, but it will
a) Falsely identify all unknown distributions as 10.14
b) Indicate "unknown distribution" for 10.14 and go hopefully smoothly on afterwards.
Afterwards, I would love to approve the changes.

**Comment by remkos** [comment]
Regarding the need for Xcode 10.1 beta 2 Command Line Tools for 10.4:
I agree that given the hassle to get the headers, we should blacklist Xcode 10.0. Seeing that Apple came around in Xcode 10.1 beta to reinstate the perl headers, that looks like the best thing to do.

Instead of waiting until Xcode 10.1 (official), which be take some time, could you alternatively just blacklist Xcode 10.0 (and maybe Xcode 10.1 beta 1), but allow Xcode 10.1 beta 2?
Installing the beta is no more trouble than installing 10.0 (except that it would not work with Xcode-select --install)

Then at least people like me (on 10.14 now) can move forward installing and developing new packages. Now I'm out dry with fink (0.43.1-121) not working because it insists it needs the Command Line Tools, even though I have them already installed.

**Comment by maxrudolph** [comment]
I also support the temporary fix suggested by remkos to allow Xcode 10.1 beta 2. I upgraded to Mojave (mea culpa) and the current state of affairs is a disaster.

**Comment by akhansen** [comment]
Feel free to contribute to the codebase and testing if you think this will be easy to accomplish.  If you don't know how to do that, then please don't editorialize.

**Comment by TheSin-** [comment]
To fix a ton of packages work with the missing /usr/include I have added this to PkgVersion.pm

```
        chomp(my $sdkpath = `xcrun --sdk macosx --show-sdk-path`);
        my %defaults = (
                "CPPFLAGS"                 => "-I\%p/include -I${sdkpath}/usr/include",
                "LDFLAGS"                  => "-L\%p/lib -L${sdkpath}/usr/lib",
        );
```

I'm not sure it's the best solution but I wanted to make record of it as it has gotten LOTS more to build cause of it"

**Comment by remkos** [comment]
@maxrudolph, it is not as bad as you say it is. For the time being, a work-around on 10.14 is to install Xcode 10.1 beta 2 and also install in addition the package /Library/Developer/CommandLineTools/Packages/macOS_SDK_headers_for_macOS_10.14.
pkg that comes with it (as explained by @jwhowarth).
This is not the prettiest solution, but it makes `fink-0.43.1-121` work again.

My comment was in the assumption that @jwhowarth had it all worked out, except for the version check on the Xcode version. Seeing @TheSin-'s addition, I guess there's more to be done.

**Comment by TheSin-** [comment]
@remkos, I'm now using the beta, I'm trying to get fink to work without having to do a version check or exception to an Xcode release, so I'm on all official releases ATM I didn't nothing but install Xcode and the cmdline tools, nothing extra, of course I"m using it with my PR and adding @jwhowarth's plus my comments above, and I"m working through the tree now to get a working php install.  apache2 and my local version of mysql-server-5.5 (in my exp) are working at 100% so far, and php-cli is now working.  So I'm getting there.

File: perlmod/Fink/PkgVersion.pm [hunk 2]
```diff
@@ -1172,7 +1172,7 @@ sub get_script {
 			# path-prefix-*wraps gcc and g++, system-perl configure hardcodes
 			# gcc-4.x, which is not wrapped or necessarily even present.
 			my ($perldirectory, $perlarchdir, $perlcmd) = $self->get_perl_dir_arch();
-			my $archflags = 'ARCHFLAGS=""'; # prevent Apple's perl from building fat
+			my $archflags = 'ARCHFLAGS="-I`xcrun --sdk macosx --show-sdk-path`/System/Library/Perl/`perl -e \'print substr($^V, 1)\' | cut -d. -f1 -f2`/darwin-thread-multi-2level/CORE"'; # prevent Apple's perl from building fat
```
[the remark this answers is not in the document]
Comment (TheSin-) [reply perlmod/Fink/PkgVersion.pm]: this is still required for Xcode 10.1

File: perlmod/Fink/VirtPackage.pm [hunk 3]
```diff
@@ -1464,10 +1468,13 @@ END
...
-	foreach my $file (qw| /usr/include |) {
-		$options{debug} && printf STDERR " - directory %s... %s\n", $file, -d $file ? "found" : "missing!";
-		$hash->{status} = STATUS_ABSENT if not -d $file;
+	my $sdkdir = '';
+	if ($osxversion >= 18) {
+		chomp($sdkdir = `xcrun --sdk macosx --show-sdk-path 2>/dev/null`);
 	}
+	my $includedir = $sdkdir . '/usr/include';
+	$options{debug} && printf STDERR " - directory %s... %s\n", $includedir, -d $includedir ? "found" : "missing!";
+	$hash->{status} = STATUS_ABSENT if not -d $includedir;
```
Comment (TheSin-) [inline perlmod/Fink/VirtPackage.pm]: This section is no longer required for Xcode 10.1 now that it is official

**Comment by remkos** [comment]
Any progress on creating a new fink release? Based on the above and the release of Xcode 10.1 (or Command Line Tools 10.1) it looks like @jwhowarth's solution is close, with small amendments from @dmacks and @TheSin-

**Comment by HyperopiaEvan** [comment]
Is there a specified package on Mojave that I can install and fix problems?

**Comment by nieder** [comment]
Is there a way to enforce Xcode 10.1 on 10.14, or are we far enough away from the 10.1 release that we can assume anyone using Mojave will now have 10.1?

**Comment by nieder** [comment]
@hyperopiaevan
````
git checkout -b jwhowarth-master master
git pull https://github.com/jwhowarth/fink.git master
````
Then run `./inject.pl' from the newly checkout out tree, which will guide you into updating full your Fink install to use the changes here.

It might need some tweaks still, though, based on the above reviews and discussions by @dmacks, @thesin-, and @remkos.

**Comment by akhansen** [comment]
We should be able to blacklist Xcode 10.0 on Mojave.

**Comment by biest3000** [comment]
Hi,

The solution in my case was to copy the complete /sw directory (16GB) 
from the operational MAC Laptop
to my MAC Desktop, junk the old /sw and replace it by the copied version.

Works like a charm but is a highly brute force and lacks any kind of 
elegance. On the other my attempts
to achieve this directly on my Desktop did cost me many hours of 
fruitless efforts!

As a humble proposal: why not offer a minimal fink installation in form 
of a sw.tgz (much less than 16GB!!!)
as an emergency back-up to overcome the lengthy transition periods. 
Maybe not not very elegant either but
should work for the same MAC OSX OS.

Cheers and many Thanks! Frank

Kevin-Gardner wrote:

**Comment by beren12** [comment]
> On Dec 2, 2018, at 6:02 PM, biest3000 <notifications@github.com> wrote:
> 
> 
> Hi,
> 
> The solution in my case was to copy the complete /sw directory (16GB) 
> from the operational MAC Laptop
> to my MAC Desktop, junk the old /sw and replace it by the copied version.

You likely have many old .deb packages and sources under /sw/src and /sw/fink/dists for /sw to be that large. 10.14 support is being worked on but between Apple’s bugs and lack of volunteers with fink it’s taking some time.

File: perlmod/Fink/Bootstrap.pm (hunk 1 above)
Comment (fingolfin) [reply perlmod/Fink/Bootstrap.pm]: Addressed in PR #170

File: perlmod/Fink/PkgVersion.pm (hunk 2 above)
[the remark this answers is not in the document]
Comment (fingolfin) [reply perlmod/Fink/PkgVersion.pm]: Ugh, this line is unreadable and IMHO unmaintainable. What is it actually doing?
First `xcrun --sdk macosx --show-sdk-path` is run to get the SDK path.
Then we run another subshell to launch perl and get the perl version -- even though we are calling from perl here. 

I think I can simplify this considerably ;-)

**Comment by akhansen** [comment]
Used #170 to achieve the same goal.

**Comment by biest3000** [comment]
Hi,

Hoping I could comment in this way again?!?

Since the initial hiccup I am happily using
fink on my MACs.

However, I have noticed 10 packages that can
no longer be updated:

- cernlib(-dev)
- extulis-makemaker-pm5182
- gcc5 (-compiler, -shlibs
- gcc6-shlibs
- gd3-shlibs
- patchy4-gfortran
- xfig

1) Some would need to temporarily uninstall
packages but would not be able to re-install
them (extulis-makemaker-pm5182, xfig)

2) Some try to install gcc8 which fails after
hours of cranking

Maybe that could be useful info for you guys.

Cheers, Frank



Alexander Hansen wrote:

**Comment by fingolfin** [comment]
@biest3000 this is a closed pull request, it won't get attention. Please open a new issue or contact the Fink support mailing list.

**Pull request closed without merging**